File: src/class01/binary_search.go
package main

import "fmt"

// 给定有序序列 arr，目标值target
// 如果arr中存在target，返回target的索引
// 如果arr中不存在target，返回二元组(l,r)，l是比target小的最大值的索引，r是比target大的最小值的索引
// 样例1：arr = [1,3,5,7]，target = 5，return 2
// 样例2：arr = [1,3,5,7]，target = 4，return (1,2)
type group struct {
	flag bool // 标识是否找到target
	l    int
	r    int
}

func binarySearch(arr []int, target int) group {
	if arr == nil {
		return group{}
	}
	l, r := 0, len(arr)-1
	mid := l + (r-l)/2
	for l <= r {
		if arr[mid] == target {
			return group{true, mid, mid}
		} else if arr[mid] > target {
			r = mid - 1
		} else {
			l = mid + 1
		}
		mid = l + (r-l)/2
	}
	// no match
	l, r = mid-1, mid
	// 边界检查
	if l < 0 {
		l = 0
	}
	if r >= len(arr) {
		r = len(arr) - 1
	}
	return group{false, l, r}
}

func search(arr []int, target int) {
	g := binarySearch(arr, target)
	if g.flag {
		fmt.Print(g.l)
	} else {
		fmt.Printf("(%v,%v)", g.l, g.r)
	}
}

// 暴力求解验证
func expected(arr []int, target int) {
	min, size := 0, len(arr)
	for i := 0; i < size; i++ {
		if arr[i] == target {
			fmt.Println(i)
			break
		}
		if arr[i] < target {
			min = i
		} else {
			break
		}
	}
	l, r := min, min+1
	if l >= size {
		r = size - 1
		l = size - 1
	}
	fmt.Printf("(%v,%v)\n", l, r)
	fmt.Printf("arr[l] = %v,arr[r] = %v", arr[l], arr[r])
}
func main() {
	arr := []int{1, 3, 5, 7}
	//arr := []int{1, 3, 5, 7, 9}
	target := -1
	search(arr, target)
	fmt.Println("\n------------------------------")
	expected(arr, target)
}

File: learn-system/src/algorithm/bisection/BinarySearch.java
package algorithm.bisection;

import java.util.Arrays;

/**
 * 二分搜索
 * @author HuanyuLee
 * @date 2023/6/11
 */
public class BinarySearch {
    /**
     * 标准的二分搜索
     * @param arr    升序数组
     * @param target 查找目标值
     * @return 目标值在arr中的索引，找不到返回-1
     */
    public static int binarySearch(int[] arr, int target) {
        if (arr == null || arr.length == 0) {
            return -1;
        }
        int l = 0;
        int r = arr.length - 1;
        while (l <= r) {
            int mid = l + ((r - l) >> 1);
            if (arr[mid] == target) {
                return mid;
            } else if (arr[mid] > target) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return -1;
    }

    /**
     * 暴力搜索
     * @param arr    升序数组
     * @param target 查找目标值
     * @return 目标值在arr中的索引，找不到返回-1
     */
    public static int search(int[] arr, int target) {
        if (arr == null || arr.length == 0) {
            return -1;
        }
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == target) {
                return i;
            }
        }
        return -1;
    }

    public static int[] generateRandomSortedArray(int maxSize, int maxValue) {
        int[] arr = new int[(int) ((maxSize + 1) * Math.random())];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = (int) ((maxValue + 1) * Math.random()) - (int) (maxValue * Math.random());
        }
        Arrays.sort(arr);
        return arr;
    }

    public static void main(String[] args) {
        int testTime = 500000;
        int maxSize = 10;
        int maxValue = 100;
        for (int i = 0; i < testTime; i++) {
            int[] arr = generateRandomSortedArray(maxSize, maxValue);
            int target = (int) ((maxValue + 1) * Math.random()) - (int) (maxValue * Math.random());
            int actual = binarySearch(arr, target);
            int expected = search(arr, target);
            if (actual != expected) {
                System.out.println("arr = " + Arrays.toString(arr));
                System.out.println("target = " + target);
                System.out.println("actual = " + actual);
                System.out.println("expected = " + expected);
                break;
            }
        }
    }
}
